azurelinuxagent/common/protocol/wire.py

```diff
@@ -1195,14 +1196,73 @@ def get_artifacts_profile(self):
                     msg = "Content: [{0}]".format(profile)
                     logger.verbose(msg)
 
-                    from azurelinuxagent.common.event import report_event, WALAEventOperation
                     report_event(op=WALAEventOperation.ArtifactsProfileBlob,
                                  is_success=False,
                                  message=msg,
                                  log_event=False)
 
         return artifacts_profile
 
+    def upload_logs(self, content):
+        try:
+            host_plugin = self.get_host_plugin()
+            host_plugin.put_vm_log(content)
+
+            msg = "Upload VM logs request succeeded using the host plugin channel for " \
+                  "container id {0} and role config file {1}".format(host_plugin.container_id,
+                                                                     host_plugin.role_config_name)
+            add_event(name=AGENT_NAME,
+                      version=CURRENT_VERSION,
+                      op=WALAEventOperation.HostPlugin,
+                      is_success=True,
+                      message=msg,
+                      log_event=True)
+        except (ResourceGoneError, InvalidContainerError) as e:
```

Comment:
If we use the AgentGlobal class here, we can avoid the extra refresh-retry because that object always has the latest guids :)

Reply:
We should try to minimize dependencies on those globals. I think the host ga plugin doesn't need that dependency since it only needs those values on initialization when an http request returns ResourceGone. Telemetry is different in that we would need  to fetch the goal state for each event in order to get the updated values; in that case the dependency on the global is a good trade  off.

Reply:
On the contrary, any HostGA request would fail if the containerId is not correct. Wouldn't it make more sense to always send the most up-to-date value we have for containerId by using AgentGlobals rather than waiting for it to fair and then retrying again. This adds a delay in the turn-around-time for the customer at the end of the day (even though minimal) so I feel we can go for that.

Also just for my understanding, why should we try to avoid dependencies on AgentGlobals? Like the way I see it is the values which always need to be up-to-date and are independent of different objects can be part of the AgentGlobals class and should be used wherever possible.

Reply:
As a general rule globals should be avoided/used sparingly. I don't buy the concern about turnaround time since the extra requests should be rare.

Reply:
Keep in mind this method would be called every hour by default, not more often. I think it makes more sense to use AgentGlobals in the telemetry code where we rely on fresh and up-to-date info instead of here.

What if the AgentGlobals happen to be out of sync here? (We just exposed a polling interval for goal state updates, so we can't rely on it being every 3 seconds and having such a narrow window.) Should this method be responsible for updating the globals? It seems cleaner to let that responsibility lie in the constructor of the goal state and let host plugin update itself and its dependencies.

Reply:
That's a very fair point, I did not think of that. Given that, I agree that we should refresh host plugin object on demand rather than using the agent global in this case.